File: Data Structures & Algorithms/three-integer-sum/submission-4.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []

        nums.sort()
        res = []
        for i in range(len(nums) - 2):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i-1]:
                continue
            target = nums[i] * -1
            j = i + 1
            k = len(nums) - 1
            while j < k:
                if nums[j] + nums[k] == target:
                    res.append([nums[i], nums[j], nums[k]])
                    k -= 1
                    j += 1
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                elif nums[j] + nums[k] < target:
                    j += 1
                else:
                    k -= 1
        return res

File: Data Structures & Algorithms/combination-target-sum-ii/submission-4.cpp
class Solution {
public:
    vector<vector<int>> res;
    void backtrack(vector<int>& nums, vector<int> cur, int target, int idx){
        if(target == 0){
            res.push_back(cur);
            return;
        }

        for(int i = idx; i < nums.size(); i++){
            if(target - nums[i] < 0)
                continue;
            if(i > idx && nums[i] == nums[i-1])
                continue;
            
            cur.push_back(nums[i]);
            backtrack(nums, cur, target - nums[i], i + 1);
            cur.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        backtrack(candidates, {}, target, 0);
        return res;
    }
};
